core/trino-main/src/main/java/io/trino/operator/scalar/JsonInputFunctions.java

```diff
@@ -0,0 +1,107 @@
+ * these input functions throw exception in case of parse error.
+ * Otherwise, the parse error is suppressed, and a marker value JSON_ERROR
+ * is returned, so that the enclosing function can handle the error accordingly
+ * to its error handling strategy (e.g. return a default value).
+ */
+public class JsonInputFunctions
+{
+    private static final ObjectMapper MAPPER = new ObjectMapper();
+    private static final Charset UTF_32LE = Charset.forName("UTF-32LE");
+
+    private JsonInputFunctions() {}
+
+    @ScalarFunction(value = "$varchar_to_json", hidden = true)
+    @SqlType(StandardTypes.JSON_2016)
+    public static JsonNode varcharToJson(@SqlType(StandardTypes.VARCHAR) Slice inputExpression, @SqlType(StandardTypes.BOOLEAN) boolean failOnError)
+    {
+        Reader reader = new InputStreamReader(inputExpression.getInput(), UTF_8);
+        return toJson(reader, failOnError);
+    }
+
+    @ScalarFunction(value = "$varbinary_to_json", hidden = true)
+    @SqlType(StandardTypes.JSON_2016)
+    public static JsonNode varbinaryToJson(@SqlType(StandardTypes.VARBINARY) Slice inputExpression, @SqlType(StandardTypes.BOOLEAN) boolean failOnError)
+    {
+        return varbinaryUtf8ToJson(inputExpression, failOnError);
+    }
+
+    @ScalarFunction(value = "$varbinary_utf8_to_json", hidden = true)
+    @SqlType(StandardTypes.JSON_2016)
+    public static JsonNode varbinaryUtf8ToJson(@SqlType(StandardTypes.VARBINARY) Slice inputExpression, @SqlType(StandardTypes.BOOLEAN) boolean failOnError)
+    {
+        Reader reader = new InputStreamReader(inputExpression.getInput(), UTF_8);
+        return toJson(reader, failOnError);
+    }
+
+    @ScalarFunction(value = "$varbinary_utf16_to_json", hidden = true)
+    @SqlType(StandardTypes.JSON_2016)
+    public static JsonNode varbinaryUtf16ToJson(@SqlType(StandardTypes.VARBINARY) Slice inputExpression, @SqlType(StandardTypes.BOOLEAN) boolean failOnError)
+    {
+        Reader reader = new InputStreamReader(inputExpression.getInput(), UTF_16LE);
+        return toJson(reader, failOnError);
+    }
+
+    @ScalarFunction(value = "$varbinary_utf32_to_json", hidden = true)
+    @SqlType(StandardTypes.JSON_2016)
+    public static JsonNode varbinaryUtf32ToJson(@SqlType(StandardTypes.VARBINARY) Slice inputExpression, @SqlType(StandardTypes.BOOLEAN) boolean failOnError)
+    {
+        Reader reader = new InputStreamReader(inputExpression.getInput(), UTF_32LE);
+        return toJson(reader, failOnError);
+    }
+
+    private static JsonNode toJson(Reader reader, boolean failOnError)
+    {
+        try {
+            return MAPPER.readTree(reader);
+        }
+        catch (JsonProcessingException e) {
+            if (failOnError) {
+                throw new JsonInputConversionError(e);
+            }
+            return JSON_ERROR;
+        }
+        catch (IOException e) {
+            throw new JsonInputConversionError(e);
```

Comment:
when is this one expected? Should it be `GENERIC_INTERNAL_ERROR` instead?

Reply:
OK, that seems reasonable. I'll propagate all IO errors which are not related to JSON itself as `GENERIC_INTERNAL_ERROR`.